mongocrypt-sys/src/test/example_state_machine.rs

```diff
@@ -0,0 +1,288 @@
+                let input =
+                    BinaryBuffer::read_json_as_bson("src/test/testdata/mongocryptd-reply.json");
+                println!(
+                    "\nmocking reply from file:\n{:?}",
+                    doc_from_binary(input.binary)
+                );
+                assert!(mongocrypt_ctx_mongo_feed(ctx, input.binary));
+                assert!(mongocrypt_ctx_mongo_done(ctx));
+            }
+            mongocrypt_ctx_state_t_MONGOCRYPT_CTX_NEED_MONGO_KEYS => {
+                let output = mongocrypt_binary_new();
+                assert!(mongocrypt_ctx_mongo_op(ctx, output));
+                println!(
+                    "\nrunning a find on the key vault coll with this filter:\n{:?}",
+                    doc_from_binary(output)
+                );
+                mongocrypt_binary_destroy(output);
+                let input = BinaryBuffer::read_json_as_bson("src/test/testdata/key-document.json");
+                println!(
+                    "\nmocking reply from file:\n{:?}",
+                    doc_from_binary(input.binary)
+                );
+                assert!(mongocrypt_ctx_mongo_feed(ctx, input.binary));
+                assert!(mongocrypt_ctx_mongo_done(ctx));
+            }
+            mongocrypt_ctx_state_t_MONGOCRYPT_CTX_NEED_KMS => {
+                loop {
+                    let kms = mongocrypt_ctx_next_kms_ctx(ctx);
+                    if kms == ptr::null_mut() {
+                        break;
+                    }
+                    let output = mongocrypt_binary_new();
+                    assert!(mongocrypt_kms_ctx_message(kms, output));
+                    with_binary_as_slice(output, |slice| {
+                        println!(
+                            "sending the following to kms:\n{:?}",
+                            str::from_utf8(slice).unwrap()
+                        );
+                    });
+                    mongocrypt_binary_destroy(output);
+                    let input = BinaryBuffer::read_http("src/test/testdata/kms-decrypt-reply.txt");
+                    println!(
+                        "mocking reply from file:\n{:?}",
+                        str::from_utf8(&input.bytes).unwrap()
+                    );
+                    assert!(mongocrypt_kms_ctx_feed(kms, input.binary));
+                    assert_eq!(0, mongocrypt_kms_ctx_bytes_needed(kms));
+                }
+                mongocrypt_ctx_kms_done(ctx);
+            }
+            mongocrypt_ctx_state_t_MONGOCRYPT_CTX_READY => {
+                let output = mongocrypt_binary_new();
+                assert!(mongocrypt_ctx_finalize(ctx, output));
+                result = doc_from_binary(output);
+                mongocrypt_binary_destroy(output);
+            }
+            mongocrypt_ctx_state_t_MONGOCRYPT_CTX_DONE => {
+                break;
+            }
+            mongocrypt_ctx_state_t_MONGOCRYPT_CTX_ERROR => {
+                let status = mongocrypt_status_new();
```

Comment:
I think we need to `mongocrypt_status_destroy` this?

as a more general comment, maybe we will want to have some CI task that checks we don't leak memory in any of the C interop code?